Created Custom Logging Class for Easy Logging

Ovec8hkin opened the pull request:
The new rsmas_logger class is an easy to use wrapper for our reasonably minimalistic logging needs that uses python's native file logging module. It is easy to setup and use and will create a unified way to handle much of our logging. For more complex logging, like that performed in `run_operations`, rsmas_logger may not be the best solution. For examples on how to use rsmas_logger, see `download_ssara_rsmas.py` and/or `download_asfserial_rsmas.py`.

Ovec8hkin committed:
Created rsmas_logger object to better organize common logging code

rsmas_logger contains routines for setting up a file and console logging handler. After initalizing the object, a special format can be specified.

Ovec8hkin committed:
Added .log method to rsmas_logging object

The log method takes a level and message and logs the message at the appropriate alert level. The behavior is identical to the normal python logging.log() method, except the level is specified with the logging.level enum rather than an integer.

Ovec8hkin committed:
Updated ssara_rsmas to use new rsmas_logging module

Ovec8hkin committed:
Updated asfserial_rsmas to use new rsmas_logging module

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+
+    format = "%(levelname)s - %(message)s"
+    console_handler, file_handler = None, None
+    logfile_name = None
```
Do you mean to make `logfile_name` a static variable? Same with the `handler`s and `logger`.

Ovec8hkin replied:
Shouldn't be static. The intention is that each script gets a separate instantiation of the rsmas_logger object with its own custom logger and handler objects, which would mean we want non-static variables.

gravelcycles replied:
Okay, then I would move those variables into the `__init__` function. Currently, they are `static` or `class variables` in Python lingo. Take a look here for more on `class` vs. `instance` variables: https://docs.python.org/2/tutorial/classes.html#class-and-instance-variables

Ovec8hkin replied:
Good to know. I wasn't aware that was how instance variables worked in python classes.

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+    logfile_name = None
+    logger = logging.getLogger()
+
+    def __init__(self, console=True, file=None):
```
`file` is a built-in function, so I would avoid using that variable name

Ovec8hkin replied:
Good catch.

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+    logger = logging.getLogger()
+
+    def __init__(self, console=True, file=None):
+
```
nit: remove whitespace when it doesn't add to readability

Ovec8hkin replied:
Which whiteface doesn't add readability: line 9 or line 11? Personally think lines above and below the method declaration do add readability.

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+        return console_handler
+
+    def set_format(self, new_format):
+
```
nit: whitespace

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+
+    def set_format(self, new_format):
+
+        if self.console_handler is not None:
```
Is this check necessary if your `console_handler` and `file_handler`s are dynamic members instead of static members of your class?

Also, more idiomatically: 
``` 
if self.console_handler:
    ...
```

Ovec8hkin replied:
Actually yes those checks can be removed.

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+
+    def setup_consolehandler(self, formatter):
+        console_handler = logging.StreamHandler()
+        console_handler.setLevel(logging.INFO)
```
Will you ever want `logging.DEBUG` for the console?

Ovec8hkin replied:
Good point. The nice thing about the logging module is that it logs everything at or higher than the given level.

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+                             "\nERROR"
+                             "\nCRITICAL")
+
+if __name__ == "__main__":
```
Let's remove these tests. Another option would be to refactor these tests into a separate file. But since these tests perform I/O, I would recommend just scrapping them after you have functionality.

Ovec8hkin replied:
No use in keeping them there to test with and to serve as an implementation example?

gravelcycles replied:
Good point. I would change the path to `file_name` in that case. Or better yet, just pipe it to STDOUT.

gravelcycles commented on download_ssara_rsmas.py:
```diff
@@ -7,149 +7,150 @@
-fileHandler = logging.FileHandler(os.getenv('OPERATIONS')+'/LOGS/ssara_rsmas.log', 'a+', encoding=None)
-fileHandler.setLevel(logging.INFO)
-fileHandler.setFormatter(std_formatter)
+# logger = logging.getLogger()
```
Commented code should be deleted.

Ovec8hkin replied:
Yep.

gravelcycles commented on download_asfserial_rsmas.py:
```diff
@@ -2,10 +2,32 @@
+logfile_name = os.getenv('OPERATIONS') + '/LOGS/asfserial_rsmas.log'
+logger = rsmas_logger(file=logfile_name)
+
+inps = None
```
I'm wary of this global variable. Can it be refactored into a local variable?

Ovec8hkin replied:
If we do this, then we have to pass it around to all of the functions. This isn't a class so their aren't really "local variables". I prefer to keep it global rather than have to pass inps as a parameter to every function call.

gravelcycles replied:
I would argue that explicitly passing `inps` to functions would make clear what those functions rely on. It could even be more clear if you only pass `inps.template` to `command_line_parse()` instead of all of `inps`. 

Besides the "_it doesn't feel right_" argument, I think the danger could arise in the case (as in this one) where `inps` ends up in an unknown (or not initiated) state and it's unclear where in the code this happens. 

Also, although this is just a script being run from the command line right now, I could see the use case arising where it is imported and run from another function. Encapsulating `inps` in that case and not exposing it seems like it could be useful.

gravelcycles replied:
Here's a complicated version where I handled `inps` in the past: https://github.com/geodesymiami/rsmas_isce/blob/master/process_sentinel.py

Ovec8hkin replied:
Not personally a fan of passing that parameter around a bunch. Seems to just complicate the argument list for otherwise simple functions, especially when its used in a minimal number of occasions across multiple functions.

But I can see where making an instance variable would be useful if we start accessing it from different scripts. However, I don't think that will ever happen, as each script should have there own inps variables holding the command line argument values for that script, and not care about the inps of other scripts.

gravelcycles replied:
I would argue that it would increase readability if function inputs are declared rather than implicit.

gravelcycles commented on download_asfserial_rsmas.py:
```diff
@@ -37,13 +59,14 @@ def run_download_asf_serial():
 
 if __name__ == "__main__":
-	
+	logger.log(logging.INFO, "DATASET: %s", str(inps.template.split('/')[-1].split(".")[0]))
```
Does this code run without error? It looks like `inps` has value `None` when this code runs.

Ovec8hkin replied:
Good catch. Might have been fixed already.

gravelcycles commented on download_ssara_rsmas.py:
```diff
@@ -7,149 +7,150 @@
-		run_ssara(run_number=run_number+1)
-
-	return 1
+    """ Runs ssara_federated_query-cj.py and checks for download issues.
```
Could you remove unchanged lines from this diff?

gravelcycles replied:
It's a bit hard to read and takes away from the important parts of the diff

Ovec8hkin replied:
How do I do this?

gravelcycles replied:
That's a good question. I've seen people do it by ammending an older commit. Sometimes they use a git IDE like SourceTree or GitX to drop specific parts of a commit

gravelcycles replied:
I wouldn't worry about it this time. I would just pay attention to this before pushing commits in general in the future. #readability woot woot!

gravelcycles replied:
Also, this makes `git blame` more useful too

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+        self.logger.addHandler(self.file_handler)
+        self.logger.addHandler(self.console_handler)
+
+    def log(self, level=logging.INFO, message="", *args, **kwargs):
```
Try refactoring this using `logger.log`: https://docs.python.org/2/library/logging.html#logging.Logger.log

Ovec8hkin replied:
Tried this. logger.log doesn't work. Im not sure why.

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,71 @@
+import logging
```
This abstraction currently isn't a subset of `logging` nor a superset of `logging`. Instead it depends on `logging` and users still need to import `logging` in order to deal with logging levels. This seems to be a messy state for this abstraction to be in.

What about adding logging levels to `rsmas_logging` (i.e. `DEBUG`, `INFO`, `WARNING`, etc)? `rsmas_logging` would then be a complete wrapper around `logging`.

Ovec8hkin replied:
I agree. Good suggestion.

gravelcycles reviewed (changes requested):
Looks pretty good overall! 

Some comments on the cleanness of the diff (whitespace, commented code, and unchanged lines picked up by `git` as changed), the `rsmas_logging` abstraction, and implementation details.

Happy to discuss any thoughts you have

Ovec8hkin committed:
Created rsmas_logging.loglevel enum to complete logging wrapper.

There is no longer a need to `import logging` as well as `import rsmas_logger`. rsmas_logger now contains an enum class that contains the same logging levels as the default logging modules and can be used instead.

Ovec8hkin committed:
Small updates and syntax changes in rsmas_logging

gravelcycles reviewed (commented):
Let's table the discussion on `global inps` until later. I think if you 1. change the `static` variables to `class` variables and 2. update the `if __name__ == "__main__":` tests, we can merge these changes.

Ovec8hkin commented:
Ok. Sounds good.

Ovec8hkin committed:
Changed static variables into instance variables

Ovec8hkin committed:
Commented out implementation tests

gravelcycles commented on rsmas_logging.py:
```diff
@@ -0,0 +1,75 @@
+                             "\nERROR"
+                             "\nCRITICAL")
+
+# if __name__ == "__main__":
```
Commented code in prod 😭.

Ovec8hkin committed:
Merge branch 'master' into logging-module
